### PR TITLE
Changes after code review

### PR DESCRIPTION
**Pull request by michalgrud**
Hi, thx for your review and time.  Referring to your Code Review comment " data is downloaded sequentially, not in parallel" : The following method "GithubApiClient.getRepositoryBranches()" already performs parallel requests for branches for given repositories in different threads, by using CompletableFuture.supplyAsync(). I could just use here Parallel stream which would be much more easier to use rather than CompletableFuture, but I tend to use the second solution at my current job to be able to manage the pool of available threads. If you want I can change that but at this moment it works fine, here are some example logs for query branches for three different repos:
     
     2023-04-25T15:44:21.491+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/GtihubProxyTestRepo/branches
     2023-04-25T15:44:21.491+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/TestAutoRUN/branches
     2023-04-25T15:44:21.491+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/GithubProxyAutorun/branches
     2023-04-25T15:44:21.491+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
     2023-04-25T15:44:21.491+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
     2023-04-25T15:44:21.491+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
     2023-04-25T15:44:21.708+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Response 200 OK
     2023-04-25T15:44:21.708+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]
     2023-04-25T15:44:21.923+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Response 200 OK
     2023-04-25T15:44:21.923+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]
     2023-04-25T15:44:21.994+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Response 200 OK
     2023-04-25T15:44:21.994+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]

public Map<String, List<GithubBranchWrapper>> getRepositoryBranches(Collection<GithubRepositoryDTO> repositoryDTOs) {

        Collection<CompletableFuture<Collection<GithubBranchWrapper>>> result = repositoryDTOs.stream()
                .map(url -> CompletableFuture.supplyAsync(() -> getRepositoryBranches(url), threadPoolService.getExecutorService()))
                .collect(Collectors.toList());

        return AsynchronousQueryUtil.getFutureQueriesResult(result, "Could not query repositories branches").stream()
                .collect(Collectors.groupingBy(GithubBranchWrapper::getRepositoryName, Collectors.toList()));
    }
`

### CODE REVIEW

File: src/main/java/com/mgrud/github/proxy/gitproxycore/domain/control/client/GithubApiClient.java [hunk 1]
```diff
@@ -18,7 +19,10 @@
 
 @Service
 public class GithubApiClient {
-    private static final String GITHUB_GET_USER_REPOSITORIES_URL = "https://api.github.com/users/{username}/repos";
+    @Value("${base.url.github}")
```
Comment (bogumilbierc) [inline src/main/java/com/mgrud/github/proxy/gitproxycore/domain/control/client/GithubApiClient.java:22]: Field based injection - can be converted to constructor injection

File: src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -44,7 +51,28 @@ public ResponseEntity<ErrorResponseDTO> handleMissingServletRequestParameterExce
         ErrorResponseDTO errorResponse = ErrorResponseDTO.builder()
                 .errorCode(ErrorResponseDTO.ErrorCodeEnum.MissingParameter)
                 .message("Parameter is missing: " + ex.getParameterName())
+                .statusCode(HttpStatus.BAD_REQUEST.value())
                 .build();
         return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
     }
+
+    @ExceptionHandler(HttpMediaTypeNotAcceptableException.class)
+    public ResponseEntity handleHttpMediaTypeNotAcceptableExceptionException(HttpMediaTypeNotAcceptableException ex) {
```
Comment (bogumilbierc) [inline src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java:60]: public ResponseEntity<ErrorResponseDTO>

File: src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -44,7 +51,28 @@ public ResponseEntity<ErrorResponseDTO> handleMissingServletRequestParameterExce
...
+
+        ErrorResponseDTO errorResponse = ErrorResponseDTO.builder()
+                .errorCode(ErrorResponseDTO.ErrorCodeEnum.NotSupportedMediaType)
+                .message("Only following media types: " + ex.getSupportedMediaTypes() + " are supported")
+                .statusCode(HttpStatus.NOT_ACCEPTABLE.value())
+                .build();
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (bogumilbierc) [inline src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java:68]: 1. You don't need object mapper here, as You can return typed response entity
2. Object Mapper is available as Bean, not need to create it
3. Try/catch with exception throwing in error handler is quite an odd pattern :)

File: src/main/resources/application-test-config.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+debug=true
+base.url.github=http://localhost:8082
```
Comment (bogumilbierc) [inline src/main/resources/application-test-config.properties:2]: What if 8082 is occupied? Wiremock should start on random port and U can refer to variable here, sth like ${wiremock.server.port} or sth similiar

File: src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java [hunk 5]
```diff
@@ -1,77 +1,204 @@
...
+    private static final String GET_REPOSITORIES_URL = "/user/repositories";
+    private static final String GET_USER_REPOSITORIES_GITHUB_API_URL = "/users/%s/repos";
+    private static final String GET_USER_REPOSITORIES_BRANCHES_GITHUB_API_URL = "/repos/%s/%s/branches";
+    private static final int PORT = 8082;
     @Autowired
     private MockMvc mvc;
 
+    private WireMockServer wireMockServer;
+
+    private ObjectMapper objectMapper = new ObjectMapper();
```
Comment (bogumilbierc) [inline src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java:62]: it is an integration test, use spring provided object mapper bean

File: src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java [hunk 6]
```diff
@@ -1,77 +1,204 @@
...
+
+
+    @Before
+    public void init() {
+        wireMockServer = new WireMockServer(new WireMockConfiguration().port(PORT));
+        wireMockServer.start();
+        WireMock.configureFor("localhost", PORT);
```
Comment (bogumilbierc) [inline src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java:69]: use random port

File: src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java [hunk 7]
```diff
@@ -1,77 +1,204 @@
...
+        stubFor(WireMock.get(urlEqualTo(String.format(GET_USER_REPOSITORIES_GITHUB_API_URL, EXISTING_USER_NAME_WITHOUT_REPOSITORIES)))
+                .willReturn(ok()
+                        .withJsonBody(objectMapper.valueToTree(Collections.emptyList()))
+                        .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)));
 
-        mvc.perform(get("/user/repositories")
-                        .contentType(MediaType.APPLICATION_JSON))
-                .andExpect(status().isBadRequest())
-                .andExpect(jsonPath("$.errorCode", is("MissingParameter")));
+        this.mvc.perform(get(GET_REPOSITORIES_URL).queryParam(USER_NAME_QUERY_PARAM, EXISTING_USER_NAME_WITHOUT_REPOSITORIES))
```
Comment (bogumilbierc) [inline src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java:79]: "this" should not be required

File: src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java [hunk 8]
```diff
@@ -1,77 +1,204 @@
...
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
+
 @RunWith(SpringRunner.class)
 @SpringBootTest
 @AutoConfigureMockMvc
+@ActiveProfiles("test-config")
```
Comment (bogumilbierc) [inline src/test/java/com/mgrud/github/proxy/gitproxycore/domain/integration/GithubProxyResourceIntegrationTest.java:41]: You don't really need a profile for this purpose. 
Simpy put test application properties into "src/test/resources"

**Comment by bogumilbierc** [comment]
> Hi, thx for your review and time. Referring to your Code Review comment " data is downloaded sequentially, not in parallel" : The following method "GithubApiClient.getRepositoryBranches()" already performs parallel requests for branches for given repositories in different threads, by using CompletableFuture.supplyAsync(). I could just use here Parallel stream which would be much more easier to use rather than CompletableFuture, but I tend to use the second solution at my current job to be able to manage the pool of available threads. If you want I can change that but at this moment it works fine, here are some example logs for query branches for three different repos:
> 
> ```
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/GtihubProxyTestRepo/branches
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/TestAutoRUN/branches
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: HTTP GET https://api.github.com/repos/Aratek666/GithubProxyAutorun/branches
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
>  2023-04-25T15:44:21.491+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Accept=[application/json, application/*+json]
>  2023-04-25T15:44:21.708+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Response 200 OK
>  2023-04-25T15:44:21.708+02:00  [y-core-thread-2] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]
>  2023-04-25T15:44:21.923+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Response 200 OK
>  2023-04-25T15:44:21.923+02:00  [y-core-thread-0] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]
>  2023-04-25T15:44:21.994+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Response 200 OK
>  2023-04-25T15:44:21.994+02:00  [y-core-thread-1] o.s.web.client.RestTemplate: Reading to [com.mgrud.github.proxy.gitproxycore.domain.control.external.dto.GithubBranchDTO[]]
> ```
> 
> public Map<String, List> getRepositoryBranches(Collection repositoryDTOs) {
> 
> ```
>     Collection<CompletableFuture<Collection<GithubBranchWrapper>>> result = repositoryDTOs.stream()
>             .map(url -> CompletableFuture.supplyAsync(() -> getRepositoryBranches(url), threadPoolService.getExecutorService()))
>             .collect(Collectors.toList());
> 
>     return AsynchronousQueryUtil.getFutureQueriesResult(result, "Could not query repositories branches").stream()
>             .collect(Collectors.groupingBy(GithubBranchWrapper::getRepositoryName, Collectors.toList()));
> }
> ```
> 
> `

Miałem chyba za duzo PRów otwartych i Twój dostał rykoszetem złego otwartego okienka.

File: src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java (hunk 2 above)
Comment (Aratek666) [reply src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java]: Fixed

File: src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java (hunk 3 above)
Comment (Aratek666) [reply src/main/java/com/mgrud/github/proxy/gitproxycore/infrastructure/exception/handler/GlobalExceptionHandler.java]: Yeee, you are totally right that was stupid what I implemented.

1) I didn't know why, but eralier when i was trying to put into Resposne.body a serializable object, it didn't return anything in the  resposne body. In this case it was only works when I was returning it as a "byte[]" or "String" and that happens only at the case with not supported MIME type in "Accept" header. But now i know why :D  Earlier i didn't define " Response.contentType()" as application/json , so yee, that was the problem which forced me to do such a stupid thing.

**Pull request merged**